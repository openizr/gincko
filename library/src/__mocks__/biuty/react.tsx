/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as React from 'react';

/**
 * biuty/react mock.
 */
function Component(props: JSXElement): JSX.Element {
  const { onChange, onFocus, onClick } = props;
  const { transform, onKeyDown, onBlur } = props;

  // Covers `onChange` handler.
  setTimeout(() => {
    onChange?.('test');
    if (process.env.IS_DATE === 'true') {
      onChange?.('2020/02/20');
    }
  }, 10);

  // Covers `onFocus` handler.
  onFocus?.();

  // Covers `onBlur` handler.
  onBlur?.();

  // Covers `transform` function.
  transform?.('ok');
  transform?.('1002');
  transform?.('100220');
  transform?.('10022020');

  // Covers `onKeyDown` handler.
  onKeyDown?.('', { key: '1' });
  onKeyDown?.('', { key: 'A', ctrlKey: false, preventDefault: vi.fn() });
  // Covers `onClick` handler.
  if (process.env.NOCLICK !== 'true') {
    setTimeout(() => {
      onClick?.();
    }, 10);
  }
  return (
    <div>{JSON.stringify(props)}</div>
  );
}

export const UILink = Component;
export const UIButton = Component;
export const UIOptions = Component;
export const UITextfield = Component;
export const UITextarea = Component;
export const UIFilePicker = Component;
export const buildClass = vi.fn((...values: string[]): string => values.join(' '));
export const markdown = vi.fn((label: string, lightMode: boolean) => `MARKDOWN FOR ${label}, ${lightMode}`);
