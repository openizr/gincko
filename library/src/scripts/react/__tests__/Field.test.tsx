/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @jest-environment jsdom
 */

import React from 'react';
import Field from 'scripts/react/Field';
import { render, waitFor } from '@testing-library/react';

jest.mock('biuty/react');
jest.mock('scripts/core/Engine');
jest.mock('scripts/react/Message');
jest.mock('scripts/react/NestedFields');

const flushPromise = jest.fn(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));

describe('react/Field', () => {
  const customComponents = {};
  const userInputs = { test: 'ok' };
  const variables = { var: 'value' };
  const i18n = jest.fn(() => 'TRANSLATED LABEL');
  const onUserAction = jest.fn() as unknown as OnUserAction;

  beforeEach(() => {
    jest.clearAllMocks();
    delete process.env.IS_DATE;
  });

  test('renders correctly - Unknown component', async () => {
    const { container } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'Unknown',
        componentProps: {},
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Message with label and helper', async () => {
    const { container } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        label: 'Test',
        status: 'success',
        component: 'Message',
        componentProps: { helper: 'Helper' },
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Message with label and message', async () => {
    const { container } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        label: 'Test',
        message: 'Success',
        status: 'success',
        component: 'Message',
        componentProps: { helper: 'Helper' },
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Button', async () => {
    const { container } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'Button',
        componentProps: {},
      }}
    />);
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    expect(onUserAction).toHaveBeenCalledTimes(1);
    expect(onUserAction).toHaveBeenCalledWith('input', 'path.0.to.field', true);
  });

  test('renders correctly - Link', async () => {
    const { container } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'Link',
        componentProps: {},
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Options', async () => {
    const { container } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'Options',
        componentProps: {
          onFocus: (): null => null,
          options: [{ label: 'test', value: 'option1' }, { value: 'option2' }],
        },
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - DatePicker', async () => {
    process.env.IS_DATE = 'true';
    const { container, rerender } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'DatePicker',
        componentProps: {},
        value: null,
      }}
    />);
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    rerender(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'DatePicker',
        value: new Date(1657192371401),
        componentProps: { placeholder: 'placeholder', debounceTimeout: 10 },
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Textfield', async () => {
    const { container, rerender } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'Textfield',
        componentProps: {},
      }}
    />);
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    rerender(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        value: 'test',
        status: 'success',
        component: 'Textfield',
        componentProps: { placeholder: 'placeholder', debounceTimeout: 10 },
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Textarea', async () => {
    const { container, rerender } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        value: null,
        status: 'success',
        component: 'Textarea',
        componentProps: {},
      }}
    />);
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    rerender(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        value: 'test',
        status: 'success',
        component: 'Textarea',
        componentProps: { placeholder: 'placeholder', debounceTimeout: 10 },
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - FilePicker', async () => {
    const { container, rerender } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'FilePicker',
        componentProps: {},
      }}
    />);
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    rerender(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'FilePicker',
        componentProps: { placeholder: 'placeholder' },
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Array', async () => {
    const { container, rerender } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'Array',
        componentProps: {},
        fields: [],
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
    rerender(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'Array',
        componentProps: {
          addButtonProps: { label: 'add' },
          removeButtonProps: { label: 'remove' },
        },
        fields: [],
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - DynamicObject', async () => {
    const { container, rerender } = render(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'DynamicObject',
        componentProps: {},
        fields: [],
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
    rerender(<Field
      isActive
      i18n={i18n}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      onUserAction={onUserAction}
      customComponents={customComponents}
      field={{
        id: 'field',
        status: 'success',
        component: 'DynamicObject',
        componentProps: {
          addTextfieldProps: { label: 'pattern', placeholder: 'placeholder' },
        },
        fields: [],
      }}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
