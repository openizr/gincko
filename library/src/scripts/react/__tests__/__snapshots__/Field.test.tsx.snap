// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`react/Field > renders correctly - Array 1`] = `
<div
  id="field"
>
  {"type":"array","path":"path.0.to.field","isActive":true,"variables":{"var":"value"},"userInputs":{"test":"ok"},"fields":[],"addButtonProps":{"label":null},"id":"path__0__to__field","removeButtonProps":{"label":null},"addTextfieldProps":{"label":null,"placeholder":null},"customComponents":{},"modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Array 2`] = `
<div
  id="field"
>
  {"type":"array","path":"path.0.to.field","isActive":true,"variables":{"var":"value"},"userInputs":{"test":"ok"},"fields":[],"addButtonProps":{"label":"TRANSLATED LABEL"},"id":"path__0__to__field","removeButtonProps":{"label":"TRANSLATED LABEL"},"addTextfieldProps":{"label":null,"placeholder":null},"customComponents":{},"modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Button 1`] = `
<div>
  {"id":"path__0__to__field","modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - DatePicker 1`] = `
<div>
  {"maxlength":10,"name":"path.0.to.field","placeholder":null,"debounceTimeout":100,"id":"path__0__to__field","readonly":false,"value":null,"modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - DatePicker 2`] = `
<div>
  {"maxlength":10,"name":"path.0.to.field","placeholder":"TRANSLATED LABEL","debounceTimeout":10,"id":"path__0__to__field","readonly":false,"value":"2022/07/07","modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - DynamicObject 1`] = `
<div
  id="field"
>
  {"type":"dynamicObject","path":"path.0.to.field","isActive":true,"variables":{"var":"value"},"userInputs":{"test":"ok"},"fields":[],"addButtonProps":{"label":null},"id":"path__0__to__field","removeButtonProps":{"label":null},"addTextfieldProps":{"label":null,"placeholder":null},"customComponents":{},"modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - DynamicObject 2`] = `
<div
  id="field"
>
  {"type":"dynamicObject","path":"path.0.to.field","isActive":true,"variables":{"var":"value"},"userInputs":{"test":"ok"},"fields":[],"addButtonProps":{"label":null},"id":"path__0__to__field","removeButtonProps":{"label":null},"addTextfieldProps":{"label":"TRANSLATED LABEL","placeholder":"TRANSLATED LABEL"},"customComponents":{},"modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - FilePicker 1`] = `
<div>
  {"name":"path.0.to.field","placeholder":null,"id":"path__0__to__field","modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - FilePicker 2`] = `
<div>
  {"name":"path.0.to.field","placeholder":"TRANSLATED LABEL","id":"path__0__to__field","modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Link 1`] = `
<div>
  {"label":"","id":"path__0__to__field","modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Message with label and helper 1`] = `
<div
  id="field"
>
  {"label":"TRANSLATED LABEL","id":"path__0__to__field","modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Message with label and message 1`] = `
<div
  id="field"
>
  {"label":"TRANSLATED LABEL","id":"path__0__to__field","modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Options 1`] = `
<div>
  {"name":"path.0.to.field","value":"undefined","options":[{"value":"option1","label":"TRANSLATED LABEL"},{"value":"option2","label":null}],"id":"path__0__to__field","modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Textarea 1`] = `
<div>
  {"name":"path.0.to.field","placeholder":null,"debounceTimeout":100,"id":"path__0__to__field","readonly":false,"value":null,"modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Textarea 2`] = `
<div>
  {"name":"path.0.to.field","placeholder":"TRANSLATED LABEL","debounceTimeout":10,"id":"path__0__to__field","readonly":false,"value":"test","modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Textfield 1`] = `
<div>
  {"name":"path.0.to.field","placeholder":null,"debounceTimeout":100,"id":"path__0__to__field","readonly":false,"modifiers":"success "}
</div>
`;

exports[`react/Field > renders correctly - Textfield 2`] = `
<div>
  {"name":"path.0.to.field","placeholder":"TRANSLATED LABEL","debounceTimeout":10,"id":"path__0__to__field","readonly":false,"modifiers":"success ","value":"test"}
</div>
`;

exports[`react/Field > renders correctly - Unknown component 1`] = `null`;
